frame/contracts/src/lib.rs

```diff
@@ -908,9 +908,45 @@ pub mod pallet {
 		StorageValue<_, DeletionQueueManager<T>, ValueQuery>;
 }
 
+/// Helper function to ensure that the origin is either a regular signed account or root
+/// and returns it as a `Caller`, if it is one of those.
+/// It errors otherwise.
+fn contract_ensure_signed_or_root<T: Config>(
+	origin: T::RuntimeOrigin,
+) -> Result<Caller<T>, BadOrigin> {
+	match ensure_signed_or_root(origin)? {
+		Some(t) => Ok(Caller::Signed(t)),
+		None => Ok(Caller::Root),
+	}
+}
+
+/// The type of callers supported by the contracts pallet.
+#[derive(Clone, Encode, Decode, PartialEq, TypeInfo, RuntimeDebugNoBound)]
+pub enum Caller<T: Config> {
```

Comment:
```suggestion
pub enum Origin<T: Config> {
```
I think "Origin" seems more appropriate here, as it's echoes the [RuntimeOrigin](https://paritytech.github.io/substrate/master/frame_support/dispatch/trait.Dispatchable.html#associatedtype.RuntimeOrigin)

Reply:
The reason is that a Caller is not necessarily the Origin, though all origins are callers. 
See this [fn](https://github.com/paritytech/substrate/pull/13708/files#diff-98cc07c06a4567de1a156035c46689e994f2436a5def2d998c91146810ba1cc4R1269)

Reply:
I see, it's a little bit mess with the terminology. *Origin* in FRAME is the extrinsic's origin. In *pallet_contracts* source code, however, it started to be used in the meaning of the *origin of the call stack*: see e.g. `caller_is_origin` host function.

Reply:
Alright, fair enough